Makefile

```diff
@@ -13,4 +13,149 @@ collector: libkindling
 .PHONY: libkindling
 libkindling:
 	mkdir -p probe/build && cd probe/build && cmake -DBUILD_DRIVER=OFF -DPROBE_VERSION=0.1.1dev .. && make 
-	mkdir -p collector/docker/libso/. && cp -rf probe/build/src/libkindling.so collector/docker/libso/. && cp -rf probe/build/src/libkindling.so /usr/lib64/.
\ No newline at end of file
+	mkdir -p collector/docker/libso/. && cp -rf probe/build/src/libkindling.so collector/docker/libso/. && cp -rf probe/build/src/libkindling.so /usr/lib64/.
+
+
+
+LIBS_SRC := $(abspath ./probe/libs/agent-libs)
+COLLECTOR_PATH := $(abspath ./collector)
+SCRIPTS_PATH := $(abspath ./deploy/scripts)
+PROBE_PATH := $(abspath ./probe)
+PROBE_PATCH_FILE := $(PROBE_PATH)/cmake/modules/agent-libs.cmake
+
+kernelVersion = $(shell uname -r)
+ErrorExit = exit 1
+
+ifneq ($(shell cat /etc/os-release | grep Ubuntu | wc -l),0)
+	KernelName = linux-headers-$(kernelVersion)
+	CheckCommand = dpkg -l | grep $(KernelName)
+	InstallCommand = sudo apt-get -y install $(KernelName)
+else
+	ifneq ($(shell cat /etc/os-release | grep CentOS | wc -l),0)
+		KernelName = kernel-devel-$(kernelVersion)
+		CheckCommand = yum list installed |grep $(KernelName)
+		InstallCommand = sudo yum -y install $(KernelName)
+	else
+		KernelName = NotSupport
+		CheckCommand = $(ErrorExit)
+		InstallCommand = $(ErrorExit)
+	endif
+endif
+
+LIBS_IMAGE ?= kindlingproject/kernel-builder:latest
+KINDLING_IMAGE ?= kindlingproject/agent-builder:latest
+
+
+## Check if probe build image and collector build image exists or pull from dockerhub
+.PHONY: builder-images
+builder-images:
+	@echo "Checking builder images...";
+	@if [ -z "$(shell docker images -q $(LIBS_IMAGE))" ]; then \
+		echo "Downloading probe build image..."; \
+		docker pull $(LIBS_IMAGE); \
+	else \
+		echo "Probe build image already exists"; \
+	fi
+	@if [ -z "$(shell docker images -q $(KINDLING_IMAGE))" ]; then \
+		echo "Downloading collector build image..."; \
+		docker pull $(KINDLING_IMAGE); \
+	else \
+		echo "Collector build image already exists"; \
+	fi
+
+
+## Check kernel headers exists or download, support Ubuntu, CentOS
+.PHONY: kernel-headers
+kernel-headers:
+	@echo "Checking kernel headers...";
+	@if [ "$(KernelName)" = "NotSupport" ]; then \
+		echo "System kernel is not the expected Ubuntu and CentOS.Try installing kernel headers locally"; \
+		$(ErrorExit); \
+	fi
+
+	@if [ -z "$(shell exit 1)" ]; then \
+		echo "Downloading $(KernelName)..."; \
+		$(InstallCommand); \
+	else \
+		echo "$(KernelName) already installed"; \
+	fi
+
+## Check libs exists or download, patch cmake file
+.PHONY: patch-libs
+patch-libs:
+	@echo "Checking libs...";
+	@if [ ! -d "$(LIBS_SRC)" ]; then \
+		echo "$(LIBS_SRC) not exist, downloading....."; \
+		git submodule update --init --recursive; \
```

Comment:
The `if` statement is always false because `agent-libs` is a directory and it's always there once you clone a fresh Kindling repository.

Reply:
Yes, your findings are valuable, and that really shouldn't be judged by `if -d $path`.

Reply:
Changed to `ls -A1q $path` to check dir is empty